libraries/standard/json/utest/json_utest.c

```diff
@@ -0,0 +1,1153 @@
+#define CUT_AFTER_ARRAY_START_MARKER             "{\"hello\": ["
+#define CUT_AFTER_ARRAY_START_MARKER_LENGTH      ( sizeof( CUT_AFTER_ARRAY_START_MARKER ) - 1 )
+
+/* Triggers the cases in which i >= max for skipObjectScalars and nextKeyValuePair. */
+#define CUT_AFTER_OBJECT_START_MARKER            "{\"hello\": {"
+#define CUT_AFTER_OBJECT_START_MARKER_LENGTH     ( sizeof( CUT_AFTER_OBJECT_START_MARKER ) - 1 )
+#define CUT_AFTER_KEY                            "{\"hello\""
+#define CUT_AFTER_KEY_LENGTH                     ( sizeof( CUT_AFTER_KEY ) - 1 )
+
+/* This prefix is used to generate multiple levels of nested objects. */
+#define NESTED_OBJECT_PREFIX                     "{\"k\":"
+#define NESTED_OBJECT_PREFIX_LENGTH              ( sizeof( NESTED_OBJECT_PREFIX ) - 1 )
+
+/* The value of the nested object with the largest depth. */
+#define NESTED_OBJECT_VALUE                      "\"v\""
+#define NESTED_OBJECT_VALUE_LENGTH               ( sizeof( NESTED_OBJECT_VALUE ) - 1 )
+
+#ifndef JSON_MAX_DEPTH
+    #define JSON_MAX_DEPTH                       32
+#endif
+
+
+/* ============================   UNITY FIXTURES ============================ */
+
+/* Called before each test method. */
+void setUp()
+{
+}
+
+/* Called after each test method. */
+void tearDown()
+{
+}
+
+/* Called at the beginning of the whole suite. */
+void suiteSetUp()
+{
+}
+
+/* Called at the end of the whole suite. */
+int suiteTearDown( int numFailures )
+{
+    return numFailures;
+}
+
+/* ========================================================================== */
+
+/**
+ * @brief Create a nested JSON array that exceeds JSON_MAX_DEPTH.
+ */
+char * allocateMaxDepthArray( void )
+{
+    size_t i, len = ( JSON_MAX_DEPTH + 1 ) * 2;
+    char * nestedArray;
+
+    nestedArray = malloc( sizeof( char ) * ( len + 1 ) );
+
+    for( i = 0; i < len / 2; i++ )
+    {
+        nestedArray[ i ] = '[';
+        nestedArray[ len - 1 - i ] = ']';
+    }
+
+    nestedArray[ len ] = '\0';
+
+    return nestedArray;
+}
+
+/**
+ * @brief Create a nested JSON object that exceeds JSON_MAX_DEPTH.
+ */
+char * allocateMaxDepthObject( void )
+{
+    size_t i = 0, len = NESTED_OBJECT_VALUE_LENGTH +
+                        ( JSON_MAX_DEPTH + 1 ) * ( NESTED_OBJECT_PREFIX_LENGTH +
+                                                   CLOSING_CURLY_BRACKET_LENGTH );
+    char * nestedObject, * nestedObjectCur;
+
+    nestedObject = malloc( sizeof( char ) * ( len + 1 ) );
+    nestedObjectCur = nestedObject;
+
+    while( i < ( JSON_MAX_DEPTH + 1 ) * NESTED_OBJECT_PREFIX_LENGTH )
+    {
+        memcpy( nestedObjectCur, NESTED_OBJECT_PREFIX, NESTED_OBJECT_PREFIX_LENGTH );
+        i += NESTED_OBJECT_PREFIX_LENGTH;
+        nestedObjectCur += NESTED_OBJECT_PREFIX_LENGTH;
+    }
+
+    memcpy( nestedObjectCur, NESTED_OBJECT_VALUE, NESTED_OBJECT_VALUE_LENGTH );
+    i += NESTED_OBJECT_VALUE_LENGTH;
+    nestedObjectCur += NESTED_OBJECT_VALUE_LENGTH;
+
+    while( i < len )
```

Comment:
Some comments about how this would result in correct number of closing brackets would be good.